charmcraft/linters.py

```diff
@@ -0,0 +1,180 @@
+
+from charmcraft.metadata import parse_metadata_yaml
+
+# type of checker/linter
+CheckType = namedtuple("CheckType", "trait warning error")(
+    trait="trait", warning="warning", error="error"
+)
+
+# result information from each checker/linter
+CheckResult = namedtuple("CheckResult", "name result url check_type text")
+
+# generic constant for the common 'unknown' result
+UNKNOWN = "unknown"
+
+# shared state between checkers, to reuse analysis results and/or other intermediate information
+shared_state = {}
+
+
+class Language:
+    """Check the language used to write the charm.
+
+    Currently only Python is detected, if the following checks are true:
+
+    - the charm has a text dispatch with a python call
+    - the charm has a `.py` entry point
+    - the entry point file is executable
+    """
+
+    check_type = CheckType.trait
+    name = "language"
+    url = "https://juju.is/docs/sdk/charmcraft-analyze#heading--language"
+    text = "The charm is written with Python."
+
+    # different result constants
+    Result = namedtuple("Result", "python unknown")(python="python", unknown=UNKNOWN)
+
+    @classmethod
+    def run(cls, basedir: pathlib.Path) -> str:
+        """Run the proper verifications."""
+        # get the entrypoint from the last useful dispatch line
+        dispatch = basedir / "dispatch"
+        entrypoint_str = ""
+        try:
+            with dispatch.open("rt", encoding="utf8") as fh:
+                last_line = None
+                for line in fh:
+                    if line.strip():
+                        last_line = line
+                if last_line:
+                    entrypoint_str = shlex.split(last_line)[-1]
+        except (IOError, UnicodeDecodeError):
+            return cls.Result.unknown
+
+        entrypoint = basedir / entrypoint_str
+        if entrypoint.suffix == ".py" and os.access(entrypoint, os.X_OK):
+            return cls.Result.python
+        return cls.Result.unknown
+
+
+class Framework:
+    """Check on which framework the charm is based on.
+
+    Currently it detects if the Operator Framework is used, if...
+
+    - the language trait is set to python
+    - the charm contains venv/ops
+    - the charm imports ops in the entry point.
+
+    ...or the Reactive Framework is used, if the charm...
+
+    - has a metadata.yaml with "name" in it
+    - has a reactive/<name>,py file that imports "charms.reactive"
+    - has a file inside "wheelhouse" dir whose name starts with "charms.reactive-"
+    """
+
+    check_type = CheckType.trait
+    name = "framework"
+    url = "https://juju.is/docs/sdk/charmcraft-analyze#heading--framework"
+    text = "The charm is based on the Operator Framework."
+
+    # different result constants
+    Result = namedtuple("Result", "operator reactive unknown")(
+        operator="operator", reactive="reactive", unknown=UNKNOWN
+    )
+
+    @staticmethod
+    def _get_imports(filepath: pathlib.Path) -> Generator[List[str], None, None]:
```

Comment:
a better name that rings imports_iterator would be nice here

Reply:
Decided to keep this.